v2/yarpctchannel/outbound.go

```diff
@@ -118,120 +114,146 @@ func callWithPeer(ctx context.Context, req *transport.Request, peer *tchannel.Pe
+	var err error
+	if o.Chooser != nil {
+		peer, onFinish, err = o.Chooser.Choose(ctx, req)
+		if err != nil {
+			return nil, nil, err
+		}
+	} else if o.Dialer != nil && o.Addr != "" {
+		peer, err = o.Dialer.RetainPeer(yarpc.Address(o.Addr), yarpc.NopSubscriber)
+		if err != nil {
+			return nil, nil, err
+		}
+		err = o.Dialer.ReleasePeer(yarpc.Address(o.Addr), yarpc.NopSubscriber)
+		if err != nil {
+			return nil, nil, err
+		}
+		onFinish = nopFinish
 	}
 
-	tp, ok := p.(*tchannelPeer)
+	tchannelPeer, ok := peer.(*tchannelPeer)
 	if !ok {
-		return nil, nil, peer.ErrInvalidPeerConversion{
-			Peer:         p,
+		return nil, nil, yarpcpeer.ErrInvalidPeerConversion{
+			Peer:         peer,
 			ExpectedType: "*tchannelPeer",
 		}
 	}
 
-	return tp, onFinish, nil
+	return tchannelPeer, onFinish, nil
 }
 
-// Transports returns the underlying TChannel Transport for this outbound.
-func (o *Outbound) Transports() []transport.Transport {
-	return []transport.Transport{o.transport}
-}
+func nopFinish(error) {}
+
+func writeBody(body io.Reader, call *tchannel.OutboundCall) error {
+	w, err := call.Arg3Writer()
+	if err != nil {
+		return err
+	}
 
-// Start starts the TChannel outbound.
-func (o *Outbound) Start() error {
-	return o.once.Start(o.chooser.Start)
+	if _, err := internaliopool.Copy(w, body); err != nil {
+		return err
+	}
+
+	return w.Close()
 }
 
-// Stop stops the TChannel outbound.
-func (o *Outbound) Stop() error {
-	return o.once.Stop(o.chooser.Stop)
+func fromSystemError(err tchannel.SystemError) error {
+	code, ok := _tchannelCodeToCode[err.Code()]
+	if !ok {
+		// This should be unreachable.
+		return yarpcerror.Newf(yarpcerror.CodeInternal, "got tchannel.SystemError %v which did not have a matching YARPC code", err)
+	}
+	return yarpcerror.Newf(code, err.Message())
 }
 
-// IsRunning returns whether the ChannelOutbound is running.
-func (o *Outbound) IsRunning() bool {
-	return o.once.IsRunning()
+// ServiceHeaderKey is internal key used by YARPC, we need to remove it before give response to client
+// only does verification when there is a response service header.
+func checkServiceMatchAndDeleteHeaderKey(reqSvcName string, resHeaders yarpc.Headers) (bool, string) {
+	if resSvcName, ok := resHeaders.Get(ServiceHeaderKey); ok {
+		resHeaders.Del(ServiceHeaderKey)
+		return reqSvcName == resSvcName, resSvcName
+	}
+	return true, ""
 }
 
-// Introspect returns basic status about this outbound.
-func (o *Outbound) Introspect() introspection.OutboundStatus {
-	state := "Stopped"
-	if o.IsRunning() {
-		state = "Running"
-	}
-	var chooser introspection.ChooserStatus
-	if i, ok := o.chooser.(introspection.IntrospectableChooser); ok {
-		chooser = i.Introspect()
-	} else {
-		chooser = introspection.ChooserStatus{
-			Name: "Introspection not available",
-		}
+func getResponseErrorAndDeleteHeaderKeys(headers yarpc.Headers) error {
+	defer func() {
+		headers.Del(ErrorCodeHeaderKey)
+		headers.Del(ErrorNameHeaderKey)
```

Comment:
nit: can avoid cost of defer on the hot path by writing a utility `popHeader` function that gets and dels at the same time.